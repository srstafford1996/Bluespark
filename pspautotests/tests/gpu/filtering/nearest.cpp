#include "shared.h"

extern "C" int HAS_DISPLAY;

typedef struct {
	float u, v;
	u32 color;
	float x, y, z;
} VertexColorF32;

static __attribute__((aligned(16))) VertexColorF32 vertices_f32[256];
static __attribute__((aligned(16))) u32 texdata[16] = {
	0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
	0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
	0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
	0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
};
static bool texdirty = true;

inline VertexColorF32 makeVertex32(float u, float v, u32 c, float x, float y, float z) {
	VertexColorF32 vv;
	vv.u = u;
	vv.v = v;
	vv.color = c;
	vv.x = x;
	vv.y = y;
	vv.z = z;
	return vv;
}

void displayBuffer(const char *reason) {
	checkpoint("%s: COLOR=%08x", reason, readDispBuffer());
	//checkpoint("%s,%d", reason, readDispBuffer() & 0xFF);

	// Reset.
	clearDispBuffer(0x44444444);
}

void drawBoxCommands(u32 c, float offsetX, float offsetY) {
	vertices_f32[0] = makeVertex32(0.0f + offsetX, 0.0f + offsetY, c, -1.0f, 1.0f, 0.0);
	vertices_f32[1] = makeVertex32(1.0f + offsetX, 1.0f + offsetY, c, -1.0f + 2 / 240.0f, 1.0f - 2 / 136.0f, 0.0);

	// Clearing cache is fun.  Let's do it all the time.
	sceKernelDcacheWritebackInvalidateRange(vertices_f32, sizeof(vertices_f32));
	sceGuDrawArray(GU_SPRITES, GU_TEXTURE_32BITF | GU_COLOR_8888 | GU_VERTEX_32BITF | GU_TRANSFORM_3D, 2, NULL, vertices_f32);
}

void init() {
	initDisplay();
	clearDispBuffer(0x44444444);
	displayBuffer("Initial");
}

void testTexLinear(const char *title, u32 texc1, u32 texc2, float offsetX = 0.0f, float offsetY = 0.0f) {
	startFrame();

	sceGuTexFilter(GU_NEAREST, GU_NEAREST);

	if (texdata[0] != texc1 || texdata[1] != texc2 || texdirty) {
		for (int i = 0; i < 16; ++i) {
			texdata[i] = (i & 1) != 0 ? texc2 : texc1;
		}
		sceKernelDcacheWritebackInvalidateRange(texdata, sizeof(texdata));

		sceGuTexImage(0, 2, 2, 4, texdata);
		sceGuTexFlush();
		sceGuTexSync();
		texdirty = false;
	}

	drawBoxCommands(0xFFFFFFFF, offsetX, offsetY);

	endFrame();

	displayBuffer(title);
}

void testTexLinearXY(const char *title, u32 texc1, u32 texc2, float offsetX = 0.0f, float offsetY = 0.0f) {
	startFrame();

	sceGuTexFilter(GU_NEAREST, GU_NEAREST);

	if (texdata[0] != texc1 || texdata[1] != texc2 || texdata[4] != texc2 || texdata[5] != texc2 || texdirty) {
		for (int i = 0; i < 16; ++i) {
			texdata[i] = i != 0 ? texc2 : texc1;
		}
		sceKernelDcacheWritebackInvalidateRange(texdata, sizeof(texdata));

		sceGuTexImage(0, 2, 2, 4, texdata);
		sceGuTexFlush();
		sceGuTexSync();
		texdirty = false;
	}

	drawBoxCommands(0xFFFFFFFF, offsetX, offsetY);

	endFrame();

	displayBuffer(title);
}

extern "C" int main(int argc, char *argv[]) {
	init();
	HAS_DISPLAY = 0;

	checkpointNext("Common:");
	testTexLinear("  Even", 0xFF0000FF, 0xFF00FF00, 0.0f);
	testTexLinear("  One texel offset", 0xFF0000FF, 0xFF00FF00, 0.5f);
	testTexLinearXY("  X and Y", 0xFF0000FF, 0xFF00FF00, 0.5f, 0.5f);
	testTexLinearXY("  Half X and Y", 0xFF0000FF, 0xFF00FF00, 0.25f, 0.25f);

	checkpointNext("Offsets:");
	static const float offsets[] = { 0.25f, 0.125f, 0.0625f, 0.03125f, 0.015625f, 0.0078125f, 0.00390625f, 0.001953125f };
	for (size_t i = 0; i < ARRAY_SIZE(offsets); ++i) {
		char temp[256];
		snprintf(temp, sizeof(temp), "  Offset %f", offsets[i]);
		testTexLinear(temp, 0xFF0000FF, 0xFF00FF00, offsets[i]);
	}
	for (int i = 0; i < 256; ++i) {
		char temp[256];
		snprintf(temp, sizeof(temp), "  Blend at %d", i);
		testTexLinear(temp, 0xFFFDFEFF, 0x00000000, (float)i / 512.0f);
	}

	checkpointNext("Half blends:");
	for (int i = 0; i < 256; ++i) {
		u32 c = 0xFF000000 | (i << 16) | (i << 8) | i;
		char temp[256];
		snprintf(temp, sizeof(temp), "  %d * half", i);
		testTexLinear(temp, c, 0x00000000, 0.25f);
	}

	checkpointNext("Pos offsets:");
	for (int i = 0; i < 16; ++i) {
		sceGuStart(GU_DIRECT, list);
		sceGuSendCommandi(76, ((2048 - (480 / 2)) << 4) + i);
		sceGuSendCommandi(77, ((2048 - (272 / 2)) << 4) + i);
		sceGuFinish();
		sceGuSync(GU_SYNC_WAIT, GU_SYNC_WHAT_DONE);

		char temp[256];
		snprintf(temp, sizeof(temp), "  Pos offset %d", i);
		testTexLinear(temp, 0xFF0000FF, 0xFF00FF00, 0.0f);
	}

	sceGuTerm();

	return 0;
}
