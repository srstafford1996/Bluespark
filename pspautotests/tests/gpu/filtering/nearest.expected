[x] Initial: COLOR=44444444

[x] Common:
[r]   Even: COLOR=440000ff
[r]   One texel offset: COLOR=4400ff00
[r]   X and Y: COLOR=4400ff00
[r]   Half X and Y: COLOR=440000ff

[x] Offsets:
[r]   Offset 0.250000: COLOR=440000ff
[r]   Offset 0.125000: COLOR=440000ff
[r]   Offset 0.062500: COLOR=440000ff
[r]   Offset 0.031250: COLOR=440000ff
[r]   Offset 0.015625: COLOR=440000ff
[r]   Offset 0.007812: COLOR=440000ff
[r]   Offset 0.003906: COLOR=440000ff
[r]   Offset 0.001953: COLOR=440000ff
[r]   Blend at 0: COLOR=44fdfeff
[r]   Blend at 1: COLOR=44fdfeff
[r]   Blend at 2: COLOR=44fdfeff
[r]   Blend at 3: COLOR=44fdfeff
[r]   Blend at 4: COLOR=44fdfeff
[r]   Blend at 5: COLOR=44fdfeff
[r]   Blend at 6: COLOR=44fdfeff
[r]   Blend at 7: COLOR=44fdfeff
[r]   Blend at 8: COLOR=44fdfeff
[r]   Blend at 9: COLOR=44fdfeff
[r]   Blend at 10: COLOR=44fdfeff
[r]   Blend at 11: COLOR=44fdfeff
[r]   Blend at 12: COLOR=44fdfeff
[r]   Blend at 13: COLOR=44fdfeff
[r]   Blend at 14: COLOR=44fdfeff
[r]   Blend at 15: COLOR=44fdfeff
[r]   Blend at 16: COLOR=44fdfeff
[r]   Blend at 17: COLOR=44fdfeff
[r]   Blend at 18: COLOR=44fdfeff
[r]   Blend at 19: COLOR=44fdfeff
[r]   Blend at 20: COLOR=44fdfeff
[r]   Blend at 21: COLOR=44fdfeff
[r]   Blend at 22: COLOR=44fdfeff
[r]   Blend at 23: COLOR=44fdfeff
[r]   Blend at 24: COLOR=44fdfeff
[r]   Blend at 25: COLOR=44fdfeff
[r]   Blend at 26: COLOR=44fdfeff
[r]   Blend at 27: COLOR=44fdfeff
[r]   Blend at 28: COLOR=44fdfeff
[r]   Blend at 29: COLOR=44fdfeff
[r]   Blend at 30: COLOR=44fdfeff
[r]   Blend at 31: COLOR=44fdfeff
[r]   Blend at 32: COLOR=44fdfeff
[r]   Blend at 33: COLOR=44fdfeff
[r]   Blend at 34: COLOR=44fdfeff
[r]   Blend at 35: COLOR=44fdfeff
[r]   Blend at 36: COLOR=44fdfeff
[r]   Blend at 37: COLOR=44fdfeff
[r]   Blend at 38: COLOR=44fdfeff
[r]   Blend at 39: COLOR=44fdfeff
[r]   Blend at 40: COLOR=44fdfeff
[r]   Blend at 41: COLOR=44fdfeff
[r]   Blend at 42: COLOR=44fdfeff
[r]   Blend at 43: COLOR=44fdfeff
[r]   Blend at 44: COLOR=44fdfeff
[r]   Blend at 45: COLOR=44fdfeff
[r]   Blend at 46: COLOR=44fdfeff
[r]   Blend at 47: COLOR=44fdfeff
[r]   Blend at 48: COLOR=44fdfeff
[r]   Blend at 49: COLOR=44fdfeff
[r]   Blend at 50: COLOR=44fdfeff
[r]   Blend at 51: COLOR=44fdfeff
[r]   Blend at 52: COLOR=44fdfeff
[r]   Blend at 53: COLOR=44fdfeff
[r]   Blend at 54: COLOR=44fdfeff
[r]   Blend at 55: COLOR=44fdfeff
[r]   Blend at 56: COLOR=44fdfeff
[r]   Blend at 57: COLOR=44fdfeff
[r]   Blend at 58: COLOR=44fdfeff
[r]   Blend at 59: COLOR=44fdfeff
[r]   Blend at 60: COLOR=44fdfeff
[r]   Blend at 61: COLOR=44fdfeff
[r]   Blend at 62: COLOR=44fdfeff
[r]   Blend at 63: COLOR=44fdfeff
[r]   Blend at 64: COLOR=44fdfeff
[r]   Blend at 65: COLOR=44fdfeff
[r]   Blend at 66: COLOR=44fdfeff
[r]   Blend at 67: COLOR=44fdfeff
[r]   Blend at 68: COLOR=44fdfeff
[r]   Blend at 69: COLOR=44fdfeff
[r]   Blend at 70: COLOR=44fdfeff
[r]   Blend at 71: COLOR=44fdfeff
[r]   Blend at 72: COLOR=44fdfeff
[r]   Blend at 73: COLOR=44fdfeff
[r]   Blend at 74: COLOR=44fdfeff
[r]   Blend at 75: COLOR=44fdfeff
[r]   Blend at 76: COLOR=44fdfeff
[r]   Blend at 77: COLOR=44fdfeff
[r]   Blend at 78: COLOR=44fdfeff
[r]   Blend at 79: COLOR=44fdfeff
[r]   Blend at 80: COLOR=44fdfeff
[r]   Blend at 81: COLOR=44fdfeff
[r]   Blend at 82: COLOR=44fdfeff
[r]   Blend at 83: COLOR=44fdfeff
[r]   Blend at 84: COLOR=44fdfeff
[r]   Blend at 85: COLOR=44fdfeff
[r]   Blend at 86: COLOR=44fdfeff
[r]   Blend at 87: COLOR=44fdfeff
[r]   Blend at 88: COLOR=44fdfeff
[r]   Blend at 89: COLOR=44fdfeff
[r]   Blend at 90: COLOR=44fdfeff
[r]   Blend at 91: COLOR=44fdfeff
[r]   Blend at 92: COLOR=44fdfeff
[r]   Blend at 93: COLOR=44fdfeff
[r]   Blend at 94: COLOR=44fdfeff
[r]   Blend at 95: COLOR=44fdfeff
[r]   Blend at 96: COLOR=44fdfeff
[r]   Blend at 97: COLOR=44fdfeff
[r]   Blend at 98: COLOR=44fdfeff
[r]   Blend at 99: COLOR=44fdfeff
[r]   Blend at 100: COLOR=44fdfeff
[r]   Blend at 101: COLOR=44fdfeff
[r]   Blend at 102: COLOR=44fdfeff
[r]   Blend at 103: COLOR=44fdfeff
[r]   Blend at 104: COLOR=44fdfeff
[r]   Blend at 105: COLOR=44fdfeff
[r]   Blend at 106: COLOR=44fdfeff
[r]   Blend at 107: COLOR=44fdfeff
[r]   Blend at 108: COLOR=44fdfeff
[r]   Blend at 109: COLOR=44fdfeff
[r]   Blend at 110: COLOR=44fdfeff
[r]   Blend at 111: COLOR=44fdfeff
[r]   Blend at 112: COLOR=44fdfeff
[r]   Blend at 113: COLOR=44fdfeff
[r]   Blend at 114: COLOR=44fdfeff
[r]   Blend at 115: COLOR=44fdfeff
[r]   Blend at 116: COLOR=44fdfeff
[r]   Blend at 117: COLOR=44fdfeff
[r]   Blend at 118: COLOR=44fdfeff
[r]   Blend at 119: COLOR=44fdfeff
[r]   Blend at 120: COLOR=44fdfeff
[r]   Blend at 121: COLOR=44fdfeff
[r]   Blend at 122: COLOR=44fdfeff
[r]   Blend at 123: COLOR=44fdfeff
[r]   Blend at 124: COLOR=44fdfeff
[r]   Blend at 125: COLOR=44fdfeff
[r]   Blend at 126: COLOR=44fdfeff
[r]   Blend at 127: COLOR=44fdfeff
[r]   Blend at 128: COLOR=44fdfeff
[r]   Blend at 129: COLOR=44fdfeff
[r]   Blend at 130: COLOR=44fdfeff
[r]   Blend at 131: COLOR=44fdfeff
[r]   Blend at 132: COLOR=44000000
[r]   Blend at 133: COLOR=44000000
[r]   Blend at 134: COLOR=44000000
[r]   Blend at 135: COLOR=44000000
[r]   Blend at 136: COLOR=44000000
[r]   Blend at 137: COLOR=44000000
[r]   Blend at 138: COLOR=44000000
[r]   Blend at 139: COLOR=44000000
[r]   Blend at 140: COLOR=44000000
[r]   Blend at 141: COLOR=44000000
[r]   Blend at 142: COLOR=44000000
[r]   Blend at 143: COLOR=44000000
[r]   Blend at 144: COLOR=44000000
[r]   Blend at 145: COLOR=44000000
[r]   Blend at 146: COLOR=44000000
[r]   Blend at 147: COLOR=44000000
[r]   Blend at 148: COLOR=44000000
[r]   Blend at 149: COLOR=44000000
[r]   Blend at 150: COLOR=44000000
[r]   Blend at 151: COLOR=44000000
[r]   Blend at 152: COLOR=44000000
[r]   Blend at 153: COLOR=44000000
[r]   Blend at 154: COLOR=44000000
[r]   Blend at 155: COLOR=44000000
[r]   Blend at 156: COLOR=44000000
[r]   Blend at 157: COLOR=44000000
[r]   Blend at 158: COLOR=44000000
[r]   Blend at 159: COLOR=44000000
[r]   Blend at 160: COLOR=44000000
[r]   Blend at 161: COLOR=44000000
[r]   Blend at 162: COLOR=44000000
[r]   Blend at 163: COLOR=44000000
[r]   Blend at 164: COLOR=44000000
[r]   Blend at 165: COLOR=44000000
[r]   Blend at 166: COLOR=44000000
[r]   Blend at 167: COLOR=44000000
[r]   Blend at 168: COLOR=44000000
[r]   Blend at 169: COLOR=44000000
[r]   Blend at 170: COLOR=44000000
[r]   Blend at 171: COLOR=44000000
[r]   Blend at 172: COLOR=44000000
[r]   Blend at 173: COLOR=44000000
[r]   Blend at 174: COLOR=44000000
[r]   Blend at 175: COLOR=44000000
[r]   Blend at 176: COLOR=44000000
[r]   Blend at 177: COLOR=44000000
[r]   Blend at 178: COLOR=44000000
[r]   Blend at 179: COLOR=44000000
[r]   Blend at 180: COLOR=44000000
[r]   Blend at 181: COLOR=44000000
[r]   Blend at 182: COLOR=44000000
[r]   Blend at 183: COLOR=44000000
[r]   Blend at 184: COLOR=44000000
[r]   Blend at 185: COLOR=44000000
[r]   Blend at 186: COLOR=44000000
[r]   Blend at 187: COLOR=44000000
[r]   Blend at 188: COLOR=44000000
[r]   Blend at 189: COLOR=44000000
[r]   Blend at 190: COLOR=44000000
[r]   Blend at 191: COLOR=44000000
[r]   Blend at 192: COLOR=44000000
[r]   Blend at 193: COLOR=44000000
[r]   Blend at 194: COLOR=44000000
[r]   Blend at 195: COLOR=44000000
[r]   Blend at 196: COLOR=44000000
[r]   Blend at 197: COLOR=44000000
[r]   Blend at 198: COLOR=44000000
[r]   Blend at 199: COLOR=44000000
[r]   Blend at 200: COLOR=44000000
[r]   Blend at 201: COLOR=44000000
[r]   Blend at 202: COLOR=44000000
[r]   Blend at 203: COLOR=44000000
[r]   Blend at 204: COLOR=44000000
[r]   Blend at 205: COLOR=44000000
[r]   Blend at 206: COLOR=44000000
[r]   Blend at 207: COLOR=44000000
[r]   Blend at 208: COLOR=44000000
[r]   Blend at 209: COLOR=44000000
[r]   Blend at 210: COLOR=44000000
[r]   Blend at 211: COLOR=44000000
[r]   Blend at 212: COLOR=44000000
[r]   Blend at 213: COLOR=44000000
[r]   Blend at 214: COLOR=44000000
[r]   Blend at 215: COLOR=44000000
[r]   Blend at 216: COLOR=44000000
[r]   Blend at 217: COLOR=44000000
[r]   Blend at 218: COLOR=44000000
[r]   Blend at 219: COLOR=44000000
[r]   Blend at 220: COLOR=44000000
[r]   Blend at 221: COLOR=44000000
[r]   Blend at 222: COLOR=44000000
[r]   Blend at 223: COLOR=44000000
[r]   Blend at 224: COLOR=44000000
[r]   Blend at 225: COLOR=44000000
[r]   Blend at 226: COLOR=44000000
[r]   Blend at 227: COLOR=44000000
[r]   Blend at 228: COLOR=44000000
[r]   Blend at 229: COLOR=44000000
[r]   Blend at 230: COLOR=44000000
[r]   Blend at 231: COLOR=44000000
[r]   Blend at 232: COLOR=44000000
[r]   Blend at 233: COLOR=44000000
[r]   Blend at 234: COLOR=44000000
[r]   Blend at 235: COLOR=44000000
[r]   Blend at 236: COLOR=44000000
[r]   Blend at 237: COLOR=44000000
[r]   Blend at 238: COLOR=44000000
[r]   Blend at 239: COLOR=44000000
[r]   Blend at 240: COLOR=44000000
[r]   Blend at 241: COLOR=44000000
[r]   Blend at 242: COLOR=44000000
[r]   Blend at 243: COLOR=44000000
[r]   Blend at 244: COLOR=44000000
[r]   Blend at 245: COLOR=44000000
[r]   Blend at 246: COLOR=44000000
[r]   Blend at 247: COLOR=44000000
[r]   Blend at 248: COLOR=44000000
[r]   Blend at 249: COLOR=44000000
[r]   Blend at 250: COLOR=44000000
[r]   Blend at 251: COLOR=44000000
[r]   Blend at 252: COLOR=44000000
[r]   Blend at 253: COLOR=44000000
[r]   Blend at 254: COLOR=44000000
[r]   Blend at 255: COLOR=44000000

[x] Half blends:
[r]   0 * half: COLOR=44000000
[r]   1 * half: COLOR=44010101
[r]   2 * half: COLOR=44020202
[r]   3 * half: COLOR=44030303
[r]   4 * half: COLOR=44040404
[r]   5 * half: COLOR=44050505
[r]   6 * half: COLOR=44060606
[r]   7 * half: COLOR=44070707
[r]   8 * half: COLOR=44080808
[r]   9 * half: COLOR=44090909
[r]   10 * half: COLOR=440a0a0a
[r]   11 * half: COLOR=440b0b0b
[r]   12 * half: COLOR=440c0c0c
[r]   13 * half: COLOR=440d0d0d
[r]   14 * half: COLOR=440e0e0e
[r]   15 * half: COLOR=440f0f0f
[r]   16 * half: COLOR=44101010
[r]   17 * half: COLOR=44111111
[r]   18 * half: COLOR=44121212
[r]   19 * half: COLOR=44131313
[r]   20 * half: COLOR=44141414
[r]   21 * half: COLOR=44151515
[r]   22 * half: COLOR=44161616
[r]   23 * half: COLOR=44171717
[r]   24 * half: COLOR=44181818
[r]   25 * half: COLOR=44191919
[r]   26 * half: COLOR=441a1a1a
[r]   27 * half: COLOR=441b1b1b
[r]   28 * half: COLOR=441c1c1c
[r]   29 * half: COLOR=441d1d1d
[r]   30 * half: COLOR=441e1e1e
[r]   31 * half: COLOR=441f1f1f
[r]   32 * half: COLOR=44202020
[r]   33 * half: COLOR=44212121
[r]   34 * half: COLOR=44222222
[r]   35 * half: COLOR=44232323
[r]   36 * half: COLOR=44242424
[r]   37 * half: COLOR=44252525
[r]   38 * half: COLOR=44262626
[r]   39 * half: COLOR=44272727
[r]   40 * half: COLOR=44282828
[r]   41 * half: COLOR=44292929
[r]   42 * half: COLOR=442a2a2a
[r]   43 * half: COLOR=442b2b2b
[r]   44 * half: COLOR=442c2c2c
[r]   45 * half: COLOR=442d2d2d
[r]   46 * half: COLOR=442e2e2e
[r]   47 * half: COLOR=442f2f2f
[r]   48 * half: COLOR=44303030
[r]   49 * half: COLOR=44313131
[r]   50 * half: COLOR=44323232
[r]   51 * half: COLOR=44333333
[r]   52 * half: COLOR=44343434
[r]   53 * half: COLOR=44353535
[r]   54 * half: COLOR=44363636
[r]   55 * half: COLOR=44373737
[r]   56 * half: COLOR=44383838
[r]   57 * half: COLOR=44393939
[r]   58 * half: COLOR=443a3a3a
[r]   59 * half: COLOR=443b3b3b
[r]   60 * half: COLOR=443c3c3c
[r]   61 * half: COLOR=443d3d3d
[r]   62 * half: COLOR=443e3e3e
[r]   63 * half: COLOR=443f3f3f
[r]   64 * half: COLOR=44404040
[r]   65 * half: COLOR=44414141
[r]   66 * half: COLOR=44424242
[r]   67 * half: COLOR=44434343
[r]   68 * half: COLOR=44444444
[r]   69 * half: COLOR=44454545
[r]   70 * half: COLOR=44464646
[r]   71 * half: COLOR=44474747
[r]   72 * half: COLOR=44484848
[r]   73 * half: COLOR=44494949
[r]   74 * half: COLOR=444a4a4a
[r]   75 * half: COLOR=444b4b4b
[r]   76 * half: COLOR=444c4c4c
[r]   77 * half: COLOR=444d4d4d
[r]   78 * half: COLOR=444e4e4e
[r]   79 * half: COLOR=444f4f4f
[r]   80 * half: COLOR=44505050
[r]   81 * half: COLOR=44515151
[r]   82 * half: COLOR=44525252
[r]   83 * half: COLOR=44535353
[r]   84 * half: COLOR=44545454
[r]   85 * half: COLOR=44555555
[r]   86 * half: COLOR=44565656
[r]   87 * half: COLOR=44575757
[r]   88 * half: COLOR=44585858
[r]   89 * half: COLOR=44595959
[r]   90 * half: COLOR=445a5a5a
[r]   91 * half: COLOR=445b5b5b
[r]   92 * half: COLOR=445c5c5c
[r]   93 * half: COLOR=445d5d5d
[r]   94 * half: COLOR=445e5e5e
[r]   95 * half: COLOR=445f5f5f
[r]   96 * half: COLOR=44606060
[r]   97 * half: COLOR=44616161
[r]   98 * half: COLOR=44626262
[r]   99 * half: COLOR=44636363
[r]   100 * half: COLOR=44646464
[r]   101 * half: COLOR=44656565
[r]   102 * half: COLOR=44666666
[r]   103 * half: COLOR=44676767
[r]   104 * half: COLOR=44686868
[r]   105 * half: COLOR=44696969
[r]   106 * half: COLOR=446a6a6a
[r]   107 * half: COLOR=446b6b6b
[r]   108 * half: COLOR=446c6c6c
[r]   109 * half: COLOR=446d6d6d
[r]   110 * half: COLOR=446e6e6e
[r]   111 * half: COLOR=446f6f6f
[r]   112 * half: COLOR=44707070
[r]   113 * half: COLOR=44717171
[r]   114 * half: COLOR=44727272
[r]   115 * half: COLOR=44737373
[r]   116 * half: COLOR=44747474
[r]   117 * half: COLOR=44757575
[r]   118 * half: COLOR=44767676
[r]   119 * half: COLOR=44777777
[r]   120 * half: COLOR=44787878
[r]   121 * half: COLOR=44797979
[r]   122 * half: COLOR=447a7a7a
[r]   123 * half: COLOR=447b7b7b
[r]   124 * half: COLOR=447c7c7c
[r]   125 * half: COLOR=447d7d7d
[r]   126 * half: COLOR=447e7e7e
[r]   127 * half: COLOR=447f7f7f
[r]   128 * half: COLOR=44808080
[r]   129 * half: COLOR=44818181
[r]   130 * half: COLOR=44828282
[r]   131 * half: COLOR=44838383
[r]   132 * half: COLOR=44848484
[r]   133 * half: COLOR=44858585
[r]   134 * half: COLOR=44868686
[r]   135 * half: COLOR=44878787
[r]   136 * half: COLOR=44888888
[r]   137 * half: COLOR=44898989
[r]   138 * half: COLOR=448a8a8a
[r]   139 * half: COLOR=448b8b8b
[r]   140 * half: COLOR=448c8c8c
[r]   141 * half: COLOR=448d8d8d
[r]   142 * half: COLOR=448e8e8e
[r]   143 * half: COLOR=448f8f8f
[r]   144 * half: COLOR=44909090
[r]   145 * half: COLOR=44919191
[r]   146 * half: COLOR=44929292
[r]   147 * half: COLOR=44939393
[r]   148 * half: COLOR=44949494
[r]   149 * half: COLOR=44959595
[r]   150 * half: COLOR=44969696
[r]   151 * half: COLOR=44979797
[r]   152 * half: COLOR=44989898
[r]   153 * half: COLOR=44999999
[r]   154 * half: COLOR=449a9a9a
[r]   155 * half: COLOR=449b9b9b
[r]   156 * half: COLOR=449c9c9c
[r]   157 * half: COLOR=449d9d9d
[r]   158 * half: COLOR=449e9e9e
[r]   159 * half: COLOR=449f9f9f
[r]   160 * half: COLOR=44a0a0a0
[r]   161 * half: COLOR=44a1a1a1
[r]   162 * half: COLOR=44a2a2a2
[r]   163 * half: COLOR=44a3a3a3
[r]   164 * half: COLOR=44a4a4a4
[r]   165 * half: COLOR=44a5a5a5
[r]   166 * half: COLOR=44a6a6a6
[r]   167 * half: COLOR=44a7a7a7
[r]   168 * half: COLOR=44a8a8a8
[r]   169 * half: COLOR=44a9a9a9
[r]   170 * half: COLOR=44aaaaaa
[r]   171 * half: COLOR=44ababab
[r]   172 * half: COLOR=44acacac
[r]   173 * half: COLOR=44adadad
[r]   174 * half: COLOR=44aeaeae
[r]   175 * half: COLOR=44afafaf
[r]   176 * half: COLOR=44b0b0b0
[r]   177 * half: COLOR=44b1b1b1
[r]   178 * half: COLOR=44b2b2b2
[r]   179 * half: COLOR=44b3b3b3
[r]   180 * half: COLOR=44b4b4b4
[r]   181 * half: COLOR=44b5b5b5
[r]   182 * half: COLOR=44b6b6b6
[r]   183 * half: COLOR=44b7b7b7
[r]   184 * half: COLOR=44b8b8b8
[r]   185 * half: COLOR=44b9b9b9
[r]   186 * half: COLOR=44bababa
[r]   187 * half: COLOR=44bbbbbb
[r]   188 * half: COLOR=44bcbcbc
[r]   189 * half: COLOR=44bdbdbd
[r]   190 * half: COLOR=44bebebe
[r]   191 * half: COLOR=44bfbfbf
[r]   192 * half: COLOR=44c0c0c0
[r]   193 * half: COLOR=44c1c1c1
[r]   194 * half: COLOR=44c2c2c2
[r]   195 * half: COLOR=44c3c3c3
[r]   196 * half: COLOR=44c4c4c4
[r]   197 * half: COLOR=44c5c5c5
[r]   198 * half: COLOR=44c6c6c6
[r]   199 * half: COLOR=44c7c7c7
[r]   200 * half: COLOR=44c8c8c8
[r]   201 * half: COLOR=44c9c9c9
[r]   202 * half: COLOR=44cacaca
[r]   203 * half: COLOR=44cbcbcb
[r]   204 * half: COLOR=44cccccc
[r]   205 * half: COLOR=44cdcdcd
[r]   206 * half: COLOR=44cecece
[r]   207 * half: COLOR=44cfcfcf
[r]   208 * half: COLOR=44d0d0d0
[r]   209 * half: COLOR=44d1d1d1
[r]   210 * half: COLOR=44d2d2d2
[r]   211 * half: COLOR=44d3d3d3
[r]   212 * half: COLOR=44d4d4d4
[r]   213 * half: COLOR=44d5d5d5
[r]   214 * half: COLOR=44d6d6d6
[r]   215 * half: COLOR=44d7d7d7
[r]   216 * half: COLOR=44d8d8d8
[r]   217 * half: COLOR=44d9d9d9
[r]   218 * half: COLOR=44dadada
[r]   219 * half: COLOR=44dbdbdb
[r]   220 * half: COLOR=44dcdcdc
[r]   221 * half: COLOR=44dddddd
[r]   222 * half: COLOR=44dedede
[r]   223 * half: COLOR=44dfdfdf
[r]   224 * half: COLOR=44e0e0e0
[r]   225 * half: COLOR=44e1e1e1
[r]   226 * half: COLOR=44e2e2e2
[r]   227 * half: COLOR=44e3e3e3
[r]   228 * half: COLOR=44e4e4e4
[r]   229 * half: COLOR=44e5e5e5
[r]   230 * half: COLOR=44e6e6e6
[r]   231 * half: COLOR=44e7e7e7
[r]   232 * half: COLOR=44e8e8e8
[r]   233 * half: COLOR=44e9e9e9
[r]   234 * half: COLOR=44eaeaea
[r]   235 * half: COLOR=44ebebeb
[r]   236 * half: COLOR=44ececec
[r]   237 * half: COLOR=44ededed
[r]   238 * half: COLOR=44eeeeee
[r]   239 * half: COLOR=44efefef
[r]   240 * half: COLOR=44f0f0f0
[r]   241 * half: COLOR=44f1f1f1
[r]   242 * half: COLOR=44f2f2f2
[r]   243 * half: COLOR=44f3f3f3
[r]   244 * half: COLOR=44f4f4f4
[r]   245 * half: COLOR=44f5f5f5
[r]   246 * half: COLOR=44f6f6f6
[r]   247 * half: COLOR=44f7f7f7
[r]   248 * half: COLOR=44f8f8f8
[r]   249 * half: COLOR=44f9f9f9
[r]   250 * half: COLOR=44fafafa
[r]   251 * half: COLOR=44fbfbfb
[r]   252 * half: COLOR=44fcfcfc
[r]   253 * half: COLOR=44fdfdfd
[r]   254 * half: COLOR=44fefefe
[r]   255 * half: COLOR=44ffffff

[x] Pos offsets:
[r]   Pos offset 0: COLOR=440000ff
[r]   Pos offset 1: COLOR=440000ff
[r]   Pos offset 2: COLOR=440000ff
[r]   Pos offset 3: COLOR=440000ff
[r]   Pos offset 4: COLOR=440000ff
[r]   Pos offset 5: COLOR=440000ff
[r]   Pos offset 6: COLOR=440000ff
[r]   Pos offset 7: COLOR=440000ff
[r]   Pos offset 8: COLOR=440000ff
[r]   Pos offset 9: COLOR=4400ff00
[r]   Pos offset 10: COLOR=4400ff00
[r]   Pos offset 11: COLOR=4400ff00
[r]   Pos offset 12: COLOR=4400ff00
[r]   Pos offset 13: COLOR=4400ff00
[r]   Pos offset 14: COLOR=4400ff00
[r]   Pos offset 15: COLOR=4400ff00
